geomstats/geometry/pre_shape.py

```diff
@@ -378,6 +379,339 @@ def integrability_tensor(self, tangent_vec_a, tangent_vec_b, base_point):
+        Michigan Mathematical Journal 13, no. 4 (December 1966): 459–69.
+        https://doi.org/10.1307/mmj/1028999604.
+        [Pennec] To appear. Equation (24) p.7.
+        """
+        hor_x = self.horizontal_projection(tangent_vec_x, base_point)
+        p_top = Matrices.transpose(base_point)
+        p_top_p = gs.matmul(p_top, base_point)
+
+        def sylv_p(mat_b):
+            return gs.linalg.solve_sylvester(
+                p_top_p, p_top_p, mat_b - Matrices.transpose(mat_b))
+
+        e_top_hor_x = gs.matmul(Matrices.transpose(tangent_vec_e), hor_x)
+        sylv_e_top_hor_x = sylv_p(e_top_hor_x)
+
+        p_top_e = gs.matmul(p_top, tangent_vec_e)
+        sylv_p_top_e = sylv_p(p_top_e)
+
+        result = gs.matmul(base_point, sylv_e_top_hor_x) \
+            + gs.matmul(hor_x, sylv_p_top_e)
+
+        return result
+
+    def integrability_tensor_derivative(
+            self, hor_tg_vec_x, hor_tg_vec_y, nabla_x_y, tg_vec_e, nabla_x_e,
+            base_point):
+        r"""Compute the covariant derivative of the integrability tensor A.
+
+        The covariant derivative :math: `\nabla_X^S (A_Y E)` is necessary to
+        compute the covariant derivative of the curvature in a submersion.
+        The components :math: `\nabla_X^S (A_Y E)` and :math: `A_Y E` are
+        computed here for the Kendall shape space at base-point
+        :math: `P = base_point` for horizontal vector fields fields :math:
+        `X, Y` extending the values :math: `X|_P = tg_vec_x`, :math: `Y|_P =
+        tg_vec_y` and a general vector field :math: `E`extending :math:
+        `E|_P = tg_vec_e` in a neighborhood of x with covariant derivatives
+        :math: `\nabla^S_X Y |_P = nabla_x_y` and
+        :math: `\nabla^S_X E |_P = nabla_x_e`.
+
+        Parameters
+        ----------
+        hor_tg_vec_x : array-like, shape=[..., {ambient_dim, [n, n]}]
+            Horizontal tangent vector at `base_point`.
+        hor_tg_vec_y : array-like, shape=[..., {ambient_dim, [n, n]}]
+            Horizontal tangent vector at `base_point`.
+        nabla_x_y : array-like, shape=[..., {ambient_dim, [n, n]}]
+            Tangent vector at `base_point`.
+        tg_vec_e : array-like, shape=[..., {ambient_dim, [n, n]}]
+            Tangent vector at `base_point`.
+        nabla_x_e : array-like, shape=[..., {ambient_dim, [n, n]}]
+            Tangent vector at `base_point`.
+        base_point : array-like, shape=[..., {ambient_dim, [n, n]}]
+            Point of the total space.
+
+        Returns
+        -------
+        nabla_x_a_y_e : array-like, shape=[..., {ambient_dim, [n, n]}]
+            Tangent vector at `base_point`, result of :math: `\nabla_X^S
+            (A_Y E)`.
+        a_y_e : array-like, shape=[..., {ambient_dim, [n, n]}]
+            Tangent vector at `base_point`, result of :math: `A_Y E`.
+
+        References
+        ----------
+        X Pennec. To be published.
+        """
+        if not gs.all(self.is_centered(base_point)):
```

Comment:
maybe use `self.belongs` to also check that the norm is one?

Reply:
Indeed